### ISSUE TITLE
Keyboard handling in iOS 9 has changed, keyboard controller no longer needed

### ISSUE THREAD

**Issue by LeoNatan** [jessesquires/JSQMessagesViewController#1063]
Hi Jesse,

As described in another issue that has been closed for some reason, the keyboard controller is not working correctly in iOS 9 betas. The issue seems to be from the keyboard controller attempting to access private view hierarchy which is no longer correct on iOS 9.

The correct design is to have the input accessory view on the controller itself, rather trying to place it by listening to frame changes. Let the system take care of it. This is how the messages app implements it. Keyboard dismiss should be done through `UIScrollView`'s `keyboardDismissMode`.

The entire premise of "keyboard controlling" is misguided, as on iOS, an application has little control of the keyboard. And it seems this mistake has been made by virtually all the messages applications as evident by the same looking bug on iOS 9.

I opened a similar issue in the past and it was dismissed as low priority because it was just related to how the input accessory view moved differently in relation to the keyboard. Now the issue is acute with iOS 9. I was half way implementing this, but I am now working on other things and will only be able to assist here and there.

Thanks,
Leo

**Comment by LeoNatan** [comment]
More for reference:

https://robots.thoughtbot.com/input-accessorizing-uiviewcontroller

**Comment by jessesquires** [comment]
Thank you sir! :+1:

**Comment by LeoNatan** [comment]
@jessesquires I have this almost completely solved. Just a matter of final touches. For example, when using an `inputAccessoryView` in a split view environment, it gets the entire frame of the keyboard (screen) instead of just the messages controller size. Thinking about a solution.

**Comment by LeoNatan** [comment]
Haha, I was afraid the solution I came up with was too messy (I use a mask to only display the input bar on the portion of the keyboard visible on the messages view controller), and see that Apple has implemented it exactly the same way:
![image1](https://cloud.githubusercontent.com/assets/2270433/8759744/d774d73c-2d04-11e5-941a-5b1bebf6eb1b.PNG)

The only problem I am seeing right now, is the way Apple handles push/pop transitions with the keyboard windows. It creates a snapshot of the keyboard windows, and the translucent toolbar does not look well. Still thinking what can be done.

Will have a PR/solution next week.

**Comment by LeoNatan** [comment]
https://medium.com/@meiwin/a-stickler-for-details-implementing-sticky-input-fields-in-ios-f88553d36dab

An interesting approach. I have managed to do it all the way, but this may be a good compromise and even smoother.

**Comment by jessesquires** [comment]
Thanks @LeoNatan ! I'm catching up on a lot of issues right now. Will try to get to this soon.

**Comment by LeoNatan** [comment]
I will push my own changes to my own branch. If you want to take a look. I have not implemented the idea above because it works quite well for me.

**Comment by AcidicSkittles** [comment]
This issue is kinda old, but I thought I would chime in. To get the keyboard to drag down and dismiss by panning there is a slight change to get the keyboardView in iOS9.

Instead of ```self.keyboardView = self.textView.inputAccessoryView.superview;```
In your ```- (void)jsq_didReceiveKeyboardDidShowNotification:(NSNotification *)notification``` method:

In iOS9 a simple workaround is 
```
if([[UIDevice currentDevice].systemVersion floatValue] >= 9.0)
{
    for(UIWindow* window in [[UIApplication sharedApplication] windows])
        if([window isKindOfClass:NSClassFromString(@"UIRemoteKeyboardWindow")])
            for(UIView* subView in window.subviews)
                if([subView isKindOfClass:NSClassFromString(@"UIInputSetHostView")])
                    for(UIView* subsubView in subView.subviews)
                        if([subsubView isKindOfClass:NSClassFromString(@"UIInputSetHostView")])
                            self.keyboardView = subsubView;
}
else
    self.keyboardView = self.inputAccessoryView.superview;
```

Based on http://stackoverflow.com/a/32821612/3831365

Seems kinda hack-ish, but I cannot find a better way

**Comment by LeoNatan** [comment]
@AcidicSkittles This is not a good way. You are using private API and you are making sure your code will break once Apple makes the smallest change. The correct way is to move the input toolbar as a `inputAccessoryView `.

**Comment by valy0** [comment]
@LeoNatan Make the input toolbar a inputAccessoryView of what?

**Comment by LeoNatan** [comment]
@valy0 The view controller.

**Comment by urte** [comment]
Guys any update on this? I have not understood if you found a way to fix this.
Thank you for your effort.

**Comment by klivin** [comment]
@AcidicSkittles I implemented your solution in my own JSQ branch and it lead to this crash when the application is dismissed:
An instance 0x13df1df80 of class UIInputSetHostView was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x13ddf0580> ( <NSKeyValueObservance 0x13ddd8d90: Observer: 0x13c7cce70, Key path: frame, Options: <New: YES, Old: YES, Prior: NO> Context: 0x10067f900, Property: 0x13df22e50> )

This occurs when users are sending messages and shut down the app immediately.  Beware using that solution, the UIInputSetHostView is released, you should probably declare the reference to the view as a strong property rather than weak, so the view doesn't go away before ready.  Haven't tested if that causes some other leak though!

**Comment by oscarvgg** [comment]
Hello, Any updates on this?

**Comment by andrew804** [comment]
When is milestone 7.2.1 or 8.0.0 which should both contain fixes for this issue likely to be released?

**Comment by mrmarcsmith** [comment]
I'm getting the same crash as @klivin.
it says "An instance 0x1381bc030 of class UIInputSetHostView was deallocated while key value observers were still registered with it. Current observation info:"

**Comment by klivin** [comment]
@mrmarcsmith You can avoid the crash by making the keyboardView a strong property of the keyboard controller as in:

@property (strong, nonatomic) UIView *keyboardView;

This fixed the crash for me.
Even so, this is still REALLY UGLY looking keyboard behavior on interactive pop/push.  The keyboard showing/hiding definitely needs some love in this library.

Any new ideas?

**Comment by sixten** [comment]
@LeoNatan is there any chance you could push your latest code? I see [your original branch](https://github.com/LeoNatan/JSQMessagesViewController/tree/develop_fix557), which is now quite a ways behind the main project, but it also sounds from your comments like you've done a bunch of work on it since then.

Also, I know that you were working on that solution before iOS 9 came out, but how far backwards compatible is it?

**Comment by jaisonv** [comment]
Is it possible to prevent this? like "turning off" the swipe down gesture to dismiss the keyboard

**Comment by kirualex** [comment]
If anybody still has some interests on this, I'm currently trying to fix this [in my fork](https://github.com/kirualex/JSQMessagesViewController).
For now I managed to remove the keyboardController altogether, handle keyboard docking properly using `inputAccessoryView` and manage the input textview via constraints only. This means the `keyboardDismissMode` can be `UIScrollViewKeyboardDismissModeInteractive`, which was the main reason I started this work.

For some reason it works perfectly when pushed on a navigation controller but doesn't show at all on a modal. Also, I've not tested on older versions yet. Any help is appreciated.

**Comment by andrew804** [comment]
@kirualex I still have interest in this as I don't know when v8 will be released. I'll take a look at your fork!

**Comment by kirualex** [comment]
Submitted a PR, but from what I see Jesse's not very active lately on the project, this may take a while. I'm pretty happy with the results, anyone is welcome to try it !

**Comment by jessesquires** [comment]
Thanks @kirualex !

**Comment by Liampronan** [comment]
was running into the same UIInputSetHostView NSKVODeallocate issue as @klivin and @mrmarcsmith 

@klivin's solution of updating the keyboardView to a strong property (from a weak property)  on JSQMessagesKeyboardController fixed it

@jessesquires and others any thoughts on this approach?

**Comment by jessesquires** [comment]
Hey all -- will be working on this soon. Trying hard to get 7.2.1 finished and released. :blush:

**Comment by eliburke** [comment]
Moved issue to milestone 8.0.0 because the fix ( #1529 ) does not work on iOS7.

**Comment by jessesquires** [comment]
@eliburke -- right, we'll take #1529 after 7.2.1, but we'll need to hack a fix for this in the meantime 😅

**Comment by jessesquires** [comment]
Closed this with a hack for now 😢 

We will merge #1529 (the proper fix) for v8.0 of the library.
